source/common/stats/heap_stat_data.cc

```diff
@@ -1,35 +1,42 @@
 #include "common/stats/heap_stat_data.h"
 
 #include "common/common/lock_guard.h"
+#include "common/common/logger.h"
 #include "common/common/thread.h"
 #include "common/common/utility.h"
 
 namespace Envoy {
 namespace Stats {
 
-HeapStatData::HeapStatData(absl::string_view key) {
-  StringUtil::strlcpy(name_, key.data(), key.size() + 1);
+HeapStatDataAllocator::~HeapStatDataAllocator() { ASSERT(stats_.empty()); }
+
+HeapStatData* HeapStatData::alloc(StatName stat_name, SymbolTable& symbol_table) {
+  void* memory = ::malloc(sizeof(HeapStatData) + stat_name.size());
+  ASSERT(memory);
+  symbol_table.incRefCount(stat_name);
+  return new (memory) HeapStatData(stat_name);
 }
 
-HeapStatDataAllocator::~HeapStatDataAllocator() { ASSERT(stats_.empty()); }
+void HeapStatData::free(SymbolTable& symbol_table) {
+  symbol_table.free(statName());
+  this->~HeapStatData();
+  ::free(this); // matches malloc() call above.
+}
 
-HeapStatData* HeapStatDataAllocator::alloc(absl::string_view name) {
-  // Any expected truncation of name is done at the callsite. No truncation is
-  // required to use this allocator. Note that data must be freed by calling
-  // its free() method, and not by destruction, thus the more complex use of
-  // unique_ptr.
-  std::unique_ptr<HeapStatData, std::function<void(HeapStatData * d)>> data(
-      HeapStatData::alloc(name), [](HeapStatData* d) { d->free(); });
+HeapStatData& HeapStatDataAllocator::alloc(StatName name) {
+  std::unique_ptr<HeapStatData, std::function<void(HeapStatData * d)>> data_ptr(
```

Comment:
Should this be `HeapStatData* d`?

Reply:
As an orthogonal note, would there be a benefit to making a named type for `std::function<void(HeapStatData*)>` which described its use here?

Reply:
RE `HeapStatData * d` I think probably the formatter made it that way, but I removed the extra space. We'll see if the formatter puts it back.

RE type-name -- this probably helps; I think it's a little more self-documenting now.